lib/oli_web/live/objectives/objectives.ex

```diff
@@ -1,431 +1,447 @@
-       objective_mappings: objective_mappings,
-       objectives_tree: objectives_tree,
-       breadcrumbs: [Breadcrumb.new(%{full_title: "Objectives"})],
-       changeset: Resources.change_revision(%Revision{}),
-       project: project,
-       subscriptions: subscriptions,
-       attachment_summary: @default_attachment_summary,
-       modal: nil,
-       author: author,
-       force_render: 0,
-       can_delete?: true,
-       edit: :none,
-       title: "Objectives | " <> project.title
-     )}
+      assign(socket,
+        breadcrumbs: [Breadcrumb.new(%{full_title: "Objectives"})],
+        project: project,
+        author: author,
+        objectives: objectives,
+        table_model: table_model,
+        total_count: length(objectives),
+        all_objectives: all_objectives,
+        all_children: all_children
+      )
+    }
   end
 
-  def render(assigns) do
-    ~L"""
-    <%= render_modal(assigns) %>
-
-    <div class="objectives container">
-      <div class="mb-2 row">
-        <div class="col-12">
-          <p>
-            Learning objectives help you to organize course content and determine appropriate assessments and instructional strategies.
-            <br/>Refer to the <a href="https://www.cmu.edu/teaching/designteach/design/learningobjectives.html" target="_blank">CMU Eberly Center guide on learning objectives</a> to learn more about the importance of attaching learning objectives to pages and activities.
-          </p>
-        </div>
-      </div>
+  defp build_objectives(project, objectives_attachments, flash_fn, first_load \\ false) do
+    all_objectives =
+      project
+      |> ObjectiveEditor.fetch_objective_mappings()
+      |> Enum.map(& &1.revision)
+
+    all_children =
+      all_objectives
+      |> Enum.reduce([], fn rev, acc -> rev.children ++ acc end)
+      |> Enum.uniq()
+
+    objectives =
+      Enum.reduce(all_objectives, [], fn rev, acc ->
+        case Enum.find(all_children, fn child_id -> child_id == rev.resource_id end) do
+          nil ->
+            mapped_children = Enum.map(rev.children, fn resource_id ->
+              Enum.find(all_objectives, fn rev -> rev.resource_id == resource_id end)
+            end)
+
+            [Map.merge(rev,
+              %{
+                children: mapped_children,
+                sub_objectives_count: length(mapped_children),
+                page_attachments_count: 0,
+                page_attachments: [],
+                activity_attachments_count: 0,
+              }
+            )] ++ acc
+          _ -> acc
+        end
+      end)
 
-      <%= live_component CreateNew, changeset: @changeset, project: @project %>
+    {:ok, table_model} = TableModel.new(objectives)
 
-      <hr class="my-4" />
+    pid = self()
+    if first_load do
+      Task.async(fn ->
+        objectives_attachments = Enum.map(objectives, fn obj ->
+          obj.resource_id
+          |> ObjectiveEditor.preview_objective_detatchment(project)
```

Comment:
Calling `preview_objective_detatchment` for each top-level objective is going to invoke a lot of database calls.  Internally this function issues 3 db calls. The chemistry course has 91 top level objectives, which means opening the Objectives view will throw 3 * 91 = 273 queries at the database.  

I would say that it is a hard requirement that this view has a constant number of database queries that get executed when it opens (regardless of whether they are synchronous within `mount` or async here in this impl).  To do that, we may have to rework how the page attachments and activity attachments are determined.  There might need to be introduced a new function that returns "page attachments" for all objectives, and one that returns "activity attachments" for all objectives - as opposed to the per objectives ones that are being used here. 

If it isn't feasible to rework some aspects of the data collection into a constant number of queries, then perhaps we can remove it from the table model, and request that data on a per objective basis when the user first clicks to expand a specific objective.

Reply:
Yeah totally agree. I just used what we had but generates a lot of queries.

Will try to write a query to retrieve page + activities attachments for all objectives and see how it goes. Thanks !